src/util/sighash.rs

```diff
@@ -0,0 +1,836 @@
+pub struct SigHashCache<T: Deref<Target = Transaction>> {
+    /// Access to transaction required for various introspection, moreover type
+    /// `T: Deref<Target=Transaction>` allows to accept borrow and mutable borrow, the
+    /// latter in particular is necessary for [SigHashCache::access_witness]
+    tx: T,
+
+    /// Common cache for taproot and segwit inputs. It's an option because it's not needed for legacy inputs
+    common_cache: Option<CommonCache>,
+
+    /// Cache for segwit v0 inputs, it's the result of another round of sha256 on `common_cache`
+    segwit_cache: Option<SegwitCache>,
+
+    /// Cache for taproot v1 inputs
+    taproot_cache: Option<TaprootCache>,
+}
+
+/// Values cached common between segwit and taproot inputs
+pub struct CommonCache {
+    prevouts: sha256::Hash,
+    sequences: sha256::Hash,
+
+    /// in theory, `outputs` could be `Option` since `NONE` and `SINGLE` doesn't need it, but since
+    /// `ALL` is the mostly used variant by large, we don't bother
+    outputs: sha256::Hash,
+}
+
+/// Values cached for segwit inputs, it's equal to [CommonCache] plus another round of `sha256`
+pub struct SegwitCache {
+    prevouts: sha256d::Hash,
+    sequences: sha256d::Hash,
+    outputs: sha256d::Hash,
+}
+
+/// Values cached for taproot inputs
+pub struct TaprootCache {
+    amounts: sha256::Hash,
+    script_pubkeys: sha256::Hash,
+}
+
+/// Contains outputs of previous transactions.
+/// In the case [SigHashType] variant is `ANYONECANPAY`, [Prevouts::One] may be provided
+pub enum Prevouts<'u> {
+    /// `One` variant allows to provide the single Prevout needed. It's useful for example
+    /// when modifier `ANYONECANPAY` is provided, only prevout of the current input is needed.
+    /// The first `usize` argument is the input index this [TxOut] is referring to.
+    One(usize, &'u TxOut),
+    /// When `ANYONECANPAY` is not provided, or the caller is handy giving all prevouts so the same
+    /// variable can be used for multiple inputs.
+    All(&'u [TxOut]),
+}
+
+const LEAF_VERSION_TAPSCRIPT: u8 = 0xc0;
+
+/// Information related to the script path spending
+pub struct ScriptPath<'s> {
+    script: &'s Script,
+    code_separator_pos: u32,
+    leaf_version: u8,
+}
+
+/// Hashtype of an input's signature, encoded in the last byte of the signature
+/// Fixed values so they can be casted as integer types for encoding
+#[derive(PartialEq, Eq, Debug, Copy, Clone)]
+pub enum SigHashType {
+    /// 0x0: Used when not explicitly specified, defaulting to [SigHashType::All]
+    Default = 0x00,
+    /// 0x1: Sign all outputs
+    All = 0x01,
+    /// 0x2: Sign no outputs --- anyone can choose the destination
+    None = 0x02,
+    /// 0x3: Sign the output whose index matches this input's index. If none exists,
+    /// sign the hash `0000000000000000000000000000000000000000000000000000000000000001`.
+    /// (This rule is probably an unintentional C++ism, but it's consensus so we have
+    /// to follow it.)
```

Comment:
Nit: we should update ths comment to say that this is only true for legacy input. For segwit (v0 and taproot), if you try to SINGLE-sign an output that doesn't exist it'll just fail.